userdocs/diagnostics/protocol-type-non-conformance.md

```diff
@@ -1,46 +1,61 @@
-}
+print(Dog.species)    // Prints "Canus familiaris"
+print(Cat.species)    // Prints "Felis catus"
+print(Animal.species) // error: static member 'species' cannot be used...
+```
 
-struct Singular: Word {
-    var word: String
-}
+Since a type conforms to a protocol only when it satisfies _all_ of that protocol's requirements, the existential type `Animal` does not conform to the protocol `Animal` because it cannot satisfy the protocol's requirement for the static property `species`:
 
-struct Plural: Word {
-    var word: String
+```swift
+func declareAnimalSpecies<T: Animal>(_ animal: T) {
+    animal.makeNoise()
+    print("My species is known as \(T.species)")
 }
 
-let singularWord = Singular(word: "mango")
-let pluralWord = Plural(word: "mangoes")
-
-let wordPairDict: [Word: Word] = [singularWord: pluralWord] // Error
+let dog = Dog()
+declareAnimalSpecies(dog)
+// Prints:
+// "Woof"
+// "My species is known as Canus familiaris"
+declareAnimalSpecies(animal)
+// error: protocol type 'Animal' cannot conform to 'Animal'...
 ```
 
-One workaround to fix this problem is to use type erasure for the protocol `Word`. Think of type erasure as a way to hide an object's type. Since `Word` is of type `Hashable`, we already have `AnyHashable` type erasure available in the standard library which we can easily use here.
+In general, any initializers, static members, and associated types required by a protocol can be used only via conforming concrete types. Although Swift allows a protocol that requires initializers or static members to be used as a type, that type _does not and cannot_ conform to the protocol itself.
 
-```swift 
-// The fix
-let wordPairDict: [AnyHashable: AnyHashable] = [singularWord: pluralWord]
-```
+Currently, even if a protocol `P` requires no initializers or static members, the existential type `P` does not conform to `P` (with exceptions below). This restriction allows library authors to add such requirements (initializers or static members) to an existing protocol without breaking their users' source code.
+
+Concrete types that _do_ conform to protocols can provide functionality similar to that of existential types. For example, the standard library provides the `AnyHashable` type for `Hashable` values. Manual implementation of such __type erasure__ can require specific knowledge of the semantic requirements for each protocol involved and is beyond the scope of this discussion.
+
+For more on using existential types, see [Protocols as Types](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html#ID275) in _The Swift Programming Language_.
 
-# Exceptions
-`@objc` protocol type with no static requirements however do conform to its own protocol. Another exception is the `Error` Swift protocol.
+## Exceptions
 
+The Swift protocol `Error` has no required members and, when used as a type, conforms to itself; `@objc` protocols with no static requirements can also be used as types that conform to themselves.
```

Comment:
`Any` is also self-conforming, though it's not technically a protocol.

Two other notes about self-conforming protocols that may be relevant:

1. By definition, self-conforming existential types satisfy the condition `P.self is P.Type`. (Although Swift's current casting implementation has bugs around this.)

2. Self-conformance is necessary for certain generic constructs in Swift to work.  You gave a basic example for generic functions, but it also applies to generic types.  It's worth talking about workarounds in this document, as this seems to be the primary reason that people run into this diagnostic:  They're trying to do perfectly reasonably things with generics that Swift in its current form does not support.

Reply:
@tbkka Thanks!

Regarding (1), are the bugs sufficiently rare that we can tell users that this relationship is true of today's Swift?

Regarding (2), can you help outline what you're referring to in terms of its applicability to generic types? Besides manual type erasure, what other workarounds are you thinking of in that case?

Reply:
1. Unfortunately, Swift 5.3 only handles this correctly for `Any`

2. Besides manual type erasure, you can avoid self-conformance problems by just using protocol-typed arguments and ivars directly, instead of making them generic:
```
// This has problems with protocol-typed arguments:
func f<T:P>(t:T) { ... }
// Instead, just use the protocol-typed argument directly:
func f(t: P) { ... }

// Similarly for generic types.
// Instead of:
struct Foo<T:P> {
    let t: T
    ....
}
// Just use a protocol-typed ivar directly and avoid the generic
struct Foo {
   let t: P
   ...
}
```
Because of how Swift handles protocols and generics, the latter is essentially the same but avoids the self-conformance problem.